downsample_cells.R

```diff
@@ -66,68 +59,82 @@ option_list = list(
+  quit(status = 2)
+}
+
+# Okay, we do have to do something, so parse matrix properly
+
+suppressPackageStartupMessages(require(DropletUtils))
+sce <- read10xCounts(opt$expression_data)
+
+metadata <- read.csv(opt$metadata, sep = "\t", stringsAsFactors = FALSE)
+
+for (field in c(opt$cell_id_field, opt$cell_type_field)){
+  if (! opt$cell_id_field %in% colnames(metadata)){
+    write(paste0("Supplied ID field: ", opt$cell_id_field, " not in metadata frame"), stderr())
+    quit(status = 1)
+  }
 }
 
-# parse the remaining data
-cell_labels = opt$cell_type_field
-cell_id_col = opt$cell_id_field
-matrix = Matrix::readMM(paste(expr_data, "matrix.mtx", sep="/"))
-metadata = read.csv(opt$metadata, sep = "\t", stringsAsFactors = FALSE)
-
-# remove technical duplicate rows
-metadata = metadata[which(!duplicated(metadata[, cell_id_col])), ]
-# get indices of all cell types; starting with most abundant
-num_per_cell_type = sort(table(metadata[, cell_labels]), decreasing = TRUE)
-grouped_indices = lapply(seq_along(num_per_cell_type),
-                   function(idx) which(metadata[, cell_labels] == names(num_per_cell_type[idx])))
-
-# build a vector of cell indices to be removed 
-total_cells_to_remove = c()
-current_cell_type_idx = 1
-
-while(current_cell_num >= cell_num_limit){
-    sample_size = floor(length(grouped_indices[[current_cell_type_idx]]) * opt$sampling_rate)
-    # sample by index, then select cells by that index
-    i = sample(1:length(grouped_indices[[current_cell_type_idx]]), sample_size)
-    current_cells_to_remove = grouped_indices[[current_cell_type_idx]][i]
-    total_cells_to_remove = append(total_cells_to_remove, current_cells_to_remove)
-
-    # remove sampled cells 
-    grouped_indices[[current_cell_type_idx]] = grouped_indices[[current_cell_type_idx]][-i]
-    # continue sampling from the same cell type or switch to next one    
-    if(current_cell_type_idx < length(num_per_cell_type)){
-        if(length(grouped_indices[[current_cell_type_idx]]) <= length(grouped_indices[[current_cell_type_idx + 1]])){
-            current_cell_type_idx = current_cell_type_idx + 1
-        }
-    } else{
-        current_cell_type_idx = 1
-    }
-    current_cell_num = current_cell_num - sample_size
+# Put the metadata into the object so we can subset both together
+colData(sce) <- merge(colData(sce), metadata, by.x='Barcode', by.y='id', all.x=TRUE, sort=FALSE)
+
+# First candidates for removal are those without a label at all
+sce <- sce[, sce[[opt$cell_type_field]] != '']
+
+# Only down-sample the most frequent cell types. Identify the ones to downsample
+# by progressively resetting the proportion of each type to that of the next
+# least abundant until total cell number falls below the limit.
+
+cell_type_freqs <- sampling_freqs <- sort(table(sce[[opt$cell_type_field]]), decreasing = TRUE)
+
+props <- cell_type_freqs/ sum(cell_type_freqs)
+classes_to_downsample <- c()
+
+for (i in 1:length(cell_type_freqs)){
+  props[1:i] <- props[i+1]
```

Comment:
I think this line needs a bit of commenting, took me a while to understand

Reply:
Added a bit- though there's also some qualitative stuff a couple of lines up